File: Perl/Problem3.pl
#! /usr/bin/perl
use strict;
use warnings;

sub isPrime {
   my $candidate = shift(@_);
   if (!($candidate % 2)) {
      return ($candidate == 2);
   }

   my $sqrRoot = sqrt($candidate);
   my $i = 3;
   while ($i <= $sqrRoot) {
      if (!($candidate % $i)) {
         return 0; #Use this for 'false' in Perl.
      }
      $i += 2;
   }

   return 1; #Use this for 'true' in Perl
}


my $number = 600851475143;
my $biggestPrime = int(sqrt($number));
if (!($biggestPrime % 2)) {
   --$biggestPrime;
}

print "Largest prime factor of $number is: ";

while ($biggestPrime > 1) {
   if (isPrime($biggestPrime) && !($number % $biggestPrime)) {
      print "$biggestPrime\n";
      $biggestPrime = 0;
   }
   $biggestPrime -= 2;
}

File: CPP/Problem3/Problem3.cpp
#include <gmpxx.h>
#include <iostream>
#include <sys/time.h>
#include <cmath>
#include "../MyTimer.h"



using namespace std;



void Problem3 ();



int main (int argc, char *argv[])
{
   timeFunc(Problem3);

   return(0);
}



/**
 * The prime factors of 13195 are 5, 7, 13 and 29. 
 *  
 * What is the largest prime factor of the number 600851475143 ? 
 */
bool isPrime (mpz_class& primeCandidate);
void Problem3 ()
{
   mpz_class bigNum;
   bigNum = "600851475143";

   mpz_class biggestPrime = sqrt(bigNum);
   cout << "Prime factors of " << bigNum << ":" << endl;
   if ((biggestPrime%2) == 0) {
      --biggestPrime;
   }
   while (biggestPrime > 1) {
      if (isPrime(biggestPrime) && ((bigNum%biggestPrime) == 0)) {
         cout << biggestPrime << endl;
         return;
      }
      biggestPrime -= 2;
   }
}
bool isPrime (mpz_class& primeCandidate)
{
   if ((primeCandidate%2) == 0) {
      return false;
   }
   mpz_class squareRoot;
   squareRoot = sqrt(primeCandidate);
   for (int i = 3; i <=  squareRoot; i+=2) {
      if ((primeCandidate%i) == 0) {
         return false;
      }
   }

   return true;
}
